File: src/domain/usecases/RepCounterFromMachine.ts
import { RepCount, RepEvent, RepType } from '../models/Models';

/**
 * Snapshot of the discovered rep ranges for UI/diagnostics.
 */
export interface RepRanges {
  minPosA: number | null;
  maxPosA: number | null;
  minPosB: number | null;
  maxPosB: number | null;
  minRangeA: [number, number] | null;
  maxRangeA: [number, number] | null;
  minRangeB: [number, number] | null;
  maxRangeB: [number, number] | null;
  rangeA: number | null;
  rangeB: number | null;
}

/**
 * Handles rep counting based on notifications emitted by the Vitruvian machine.
 *
 * This is a direct port of the logic used by the reference web application. Rather than trying to
 * infer reps from position data, we track the counters supplied by the hardware (u16 values) and
 * supplement them with light position tracking for range calibration and auto-stop support.
 */
export class RepCounterFromMachine {
  private warmupReps: number = 0;
  private workingReps: number = 0;
  private warmupTarget: number = 3;
  private workingTarget: number = 0;
  private isJustLift: boolean = false;
  private stopAtTop: boolean = false;
  private shouldStop: boolean = false;

  private lastTopCounter: number | null = null;
  private lastCompleteCounter: number | null = null;

  private topPositionsA: number[] = [];
  private topPositionsB: number[] = [];
  private bottomPositionsA: number[] = [];
  private bottomPositionsB: number[] = [];

  private maxRepPosA: number | null = null;
  private minRepPosA: number | null = null;
  private maxRepPosB: number | null = null;
  private minRepPosB: number | null = null;

  private maxRepPosARange: [number, number] | null = null;
  private minRepPosARange: [number, number] | null = null;
  private maxRepPosBRange: [number, number] | null = null;
  private minRepPosBRange: [number, number] | null = null;

  public onRepEvent: ((event: RepEvent) => void) | null = null;

  /**
   * Configure the rep counter with workout parameters
   */
  public configure(
    warmupTarget: number,
    workingTarget: number,
    isJustLift: boolean,
    stopAtTop: boolean
  ): void {
    this.warmupTarget = warmupTarget;
    this.workingTarget = workingTarget;
    this.isJustLift = isJustLift;
    this.stopAtTop = stopAtTop;
  }

  /**
   * Reset the rep counter to initial state
   */
  public reset(): void {
    this.warmupReps = 0;
    this.workingReps = 0;
    this.shouldStop = false;
    this.lastTopCounter = null;
    this.lastCompleteCounter = null;
    this.topPositionsA = [];
    this.topPositionsB = [];
    this.bottomPositionsA = [];
    this.bottomPositionsB = [];
    this.maxRepPosA = null;
    this.minRepPosA = null;
    this.maxRepPosB = null;
    this.minRepPosB = null;
    this.maxRepPosARange = null;
    this.minRepPosARange = null;
    this.maxRepPosBRange = null;
    this.minRepPosBRange = null;
  }

  /**
   * Process a new counter update from the machine
   */
  public process(topCounter: number, completeCounter: number, posA: number = 0, posB: number = 0): void {
    // Handle top counter
    if (this.lastTopCounter !== null) {
      const topDelta = this.calculateDelta(this.lastTopCounter, topCounter);
      if (topDelta > 0) {
        this.recordTopPosition(posA, posB);

        // SAFETY FIX: Stop at top position AFTER completing target reps
        // This ensures full tension through both concentric and eccentric of final rep
        // Changed from (workingTarget - 1) to (workingTarget) so we don't release early
        if (
          this.stopAtTop &&
          !this.isJustLift &&
          this.workingTarget > 0 &&
          this.workingReps >= this.workingTarget
        ) {
          this.shouldStop = true;
          this.onRepEvent?.({
            type: RepType.WORKOUT_COMPLETE,
            warmupCount: this.warmupReps,
            workingCount: this.workingReps,
            timestamp: Date.now(),
          });
        }
      }
    }
    this.lastTopCounter = topCounter;

    // Handle complete counter
    if (this.lastCompleteCounter === null) {
      this.lastCompleteCounter = completeCounter;
      return; // Skip first signal to establish baseline
    }

    const delta = this.calculateDelta(this.lastCompleteCounter, completeCounter);
    if (delta <= 0) {
      return;
    }

    this.lastCompleteCounter = completeCounter;

    this.recordBottomPosition(posA, posB);

    const totalReps = this.warmupReps + this.workingReps + 1;
    if (totalReps <= this.warmupTarget) {
      // Warmup rep
      this.warmupReps++;
      this.onRepEvent?.({
        type: RepType.WARMUP_COMPLETED,
        warmupCount: this.warmupReps,
        workingCount: this.workingReps,
        timestamp: Date.now(),
      });

      if (this.warmupReps === this.warmupTarget) {
        this.onRepEvent?.({
          type: RepType.WARMUP_COMPLETE,
          warmupCount: this.warmupReps,
          workingCount: this.workingReps,
          timestamp: Date.now(),
        });
      }
    } else {
      // Working rep
      this.workingReps++;
      this.onRepEvent?.({
        type: RepType.WORKING_COMPLETED,
        warmupCount: this.warmupReps,
        workingCount: this.workingReps,
        timestamp: Date.now(),
      });

      // Only stop at bottom if stopAtTop is disabled
      // Most users should use stopAtTop (safer) but this preserves old behavior for those who want it
      if (
        !this.stopAtTop &&
        !this.isJustLift &&
        this.workingTarget > 0 &&
        this.workingReps >= this.workingTarget
      ) {
        this.shouldStop = true;
        this.onRepEvent?.({
          type: RepType.WORKOUT_COMPLETE,
          warmupCount: this.warmupReps,
          workingCount: this.workingReps,
          timestamp: Date.now(),
        });
      }
    }
  }

  /**
   * Calculate delta between two u16 counter values, handling overflow
   */
  private calculateDelta(last: number, current: number): number {
    if (current >= last) {
      return current - last;
    } else {
      // Handle u16 overflow
      return 0xffff - last + current + 1;
    }
  }

  /**
   * Record a top position measurement
   */
  private recordTopPosition(posA: number, posB: number): void {
    if (posA <= 0 && posB <= 0) return;

    const window = this.getWindowSize();
    if (posA > 0) {
      this.topPositionsA.push(posA);
      if (this.topPositionsA.length > window) {
        this.topPositionsA.shift();
      }
    }
    if (posB > 0) {
      this.topPositionsB.push(posB);
      if (this.topPositionsB.length > window) {
        this.topPositionsB.shift();
      }
    }

    this.updateRepRanges();
  }

  /**
   * Record a bottom position measurement
   */
  private recordBottomPosition(posA: number, posB: number): void {
    if (posA <= 0 && posB <= 0) return;

    const window = this.getWindowSize();
    if (posA > 0) {
      this.bottomPositionsA.push(posA);
      if (this.bottomPositionsA.length > window) {
        this.bottomPositionsA.shift();
      }
    }
    if (posB > 0) {
      this.bottomPositionsB.push(posB);
      if (this.bottomPositionsB.length > window) {
        this.bottomPositionsB.shift();
      }
    }

    this.updateRepRanges();
  }

  /**
   * Update the computed rep ranges from position samples
   */
  private updateRepRanges(): void {
    if (this.topPositionsA.length > 0) {
      this.maxRepPosA = Math.round(
        this.topPositionsA.reduce((sum, val) => sum + val, 0) / this.topPositionsA.length
      );
      this.maxRepPosARange = [Math.min(...this.topPositionsA), Math.max(...this.topPositionsA)];
    }
    if (this.bottomPositionsA.length > 0) {
      this.minRepPosA = Math.round(
        this.bottomPositionsA.reduce((sum, val) => sum + val, 0) / this.bottomPositionsA.length
      );
      this.minRepPosARange = [Math.min(...this.bottomPositionsA), Math.max(...this.bottomPositionsA)];
    }
    if (this.topPositionsB.length > 0) {
      this.maxRepPosB = Math.round(
        this.topPositionsB.reduce((sum, val) => sum + val, 0) / this.topPositionsB.length
      );
      this.maxRepPosBRange = [Math.min(...this.topPositionsB), Math.max(...this.topPositionsB)];
    }
    if (this.bottomPositionsB.length > 0) {
      this.minRepPosB = Math.round(
        this.bottomPositionsB.reduce((sum, val) => sum + val, 0) / this.bottomPositionsB.length
      );
      this.minRepPosBRange = [Math.min(...this.bottomPositionsB), Math.max(...this.bottomPositionsB)];
    }
  }

  /**
   * Get the sliding window size for position averaging
   */
  private getWindowSize(): number {
    const total = this.warmupReps + this.workingReps;
    return total < this.warmupTarget ? 2 : 3;
  }

  /**
   * Get the current rep count
   */
  public getRepCount(): RepCount {
    const total = this.workingReps; // Exclude warm-up reps from total count
    return {
      warmupReps: this.warmupReps,
      workingReps: this.workingReps,
      totalReps: total,
      isWarmupComplete: this.warmupReps >= this.warmupTarget,
    };
  }

  /**
   * Check if the workout should stop
   */
  public shouldStopWorkout(): boolean {
    return this.shouldStop;
  }

  /**
   * Get the calibrated top position
   */
  public getCalibratedTopPosition(): number | null {
    return this.maxRepPosA;
  }

  /**
   * Get the current rep ranges
   */
  public getRepRanges(): RepRanges {
    const rangeA =
      this.minRepPosA !== null && this.maxRepPosA !== null
        ? Math.max(this.maxRepPosA - this.minRepPosA, 0)
        : null;
    const rangeB =
      this.minRepPosB !== null && this.maxRepPosB !== null
        ? Math.max(this.maxRepPosB - this.minRepPosB, 0)
        : null;

    return {
      minPosA: this.minRepPosA,
      maxPosA: this.maxRepPosA,
      minPosB: this.minRepPosB,
      maxPosB: this.maxRepPosB,
      minRangeA: this.minRepPosARange,
      maxRangeA: this.maxRepPosARange,
      minRangeB: this.minRepPosBRange,
      maxRangeB: this.maxRepPosBRange,
      rangeA,
      rangeB,
    };
  }

  /**
   * Check if we have a meaningful range of motion
   */
  public hasMeaningfulRange(minRangeThreshold: number = 50): boolean {
    const rangeA =
      this.minRepPosA !== null && this.maxRepPosA !== null ? this.maxRepPosA - this.minRepPosA : 0;
    const rangeB =
      this.minRepPosB !== null && this.maxRepPosB !== null ? this.maxRepPosB - this.minRepPosB : 0;
    return rangeA > minRangeThreshold || rangeB > minRangeThreshold;
  }

  /**
   * Check if current position is in the danger zone (near bottom)
   */
  public isInDangerZone(posA: number, posB: number, minRangeThreshold: number = 50): boolean {
    const checkA = this.minRepPosA !== null && this.maxRepPosA !== null;
    const checkB = this.minRepPosB !== null && this.maxRepPosB !== null;
    if (!checkA && !checkB) return false;

    let danger = false;
    if (checkA) {
      const rangeA = this.maxRepPosA! - this.minRepPosA!;
      if (rangeA > minRangeThreshold) {
        const thresholdA = this.minRepPosA! + Math.floor(rangeA * 0.05);
        danger = danger || posA <= thresholdA;
      }
    }
    if (checkB) {
      const rangeB = this.maxRepPosB! - this.minRepPosB!;
      if (rangeB > minRangeThreshold) {
        const thresholdB = this.minRepPosB! + Math.floor(rangeB * 0.05);
        danger = danger || posB <= thresholdB;
      }
    }
    return danger;
  }
}

File: deprecated/app/src/main/java/com/example/vitruvianredux/domain/usecase/RepCounterFromMachine.kt
package com.example.vitruvianredux.domain.usecase

import com.example.vitruvianredux.domain.model.RepCount
import com.example.vitruvianredux.domain.model.RepEvent
import com.example.vitruvianredux.domain.model.RepType
import kotlin.math.max

/**
 * Handles rep counting based on notifications emitted by the Vitruvian machine.
 *
 * This is a direct port of the logic used by the reference web application. Rather than trying to
 * infer reps from position data, we track the counters supplied by the hardware (u16 values) and
 * supplement them with light position tracking for range calibration and auto-stop support.
 */
class RepCounterFromMachine {

    private var warmupReps = 0
    private var workingReps = 0
    private var warmupTarget = 3
    private var workingTarget = 0
    private var isJustLift = false
    private var stopAtTop = false
    private var shouldStop = false

    private var lastTopCounter: Int? = null
    private var lastCompleteCounter: Int? = null

    private val topPositionsA = mutableListOf<Int>()
    private val topPositionsB = mutableListOf<Int>()
    private val bottomPositionsA = mutableListOf<Int>()
    private val bottomPositionsB = mutableListOf<Int>()

    private var maxRepPosA: Int? = null
    private var minRepPosA: Int? = null
    private var maxRepPosB: Int? = null
    private var minRepPosB: Int? = null

    private var maxRepPosARange: Pair<Int, Int>? = null
    private var minRepPosARange: Pair<Int, Int>? = null
    private var maxRepPosBRange: Pair<Int, Int>? = null
    private var minRepPosBRange: Pair<Int, Int>? = null

    var onRepEvent: ((RepEvent) -> Unit)? = null

    fun configure(
        warmupTarget: Int,
        workingTarget: Int,
        isJustLift: Boolean,
        stopAtTop: Boolean
    ) {
        this.warmupTarget = warmupTarget
        this.workingTarget = workingTarget
        this.isJustLift = isJustLift
        this.stopAtTop = stopAtTop
    }

    fun reset() {
        warmupReps = 0
        workingReps = 0
        shouldStop = false
        lastTopCounter = null
        lastCompleteCounter = null
        topPositionsA.clear()
        topPositionsB.clear()
        bottomPositionsA.clear()
        bottomPositionsB.clear()
        maxRepPosA = null
        minRepPosA = null
        maxRepPosB = null
        minRepPosB = null
        maxRepPosARange = null
        minRepPosARange = null
        maxRepPosBRange = null
        minRepPosBRange = null
    }

    fun process(topCounter: Int, completeCounter: Int, posA: Int = 0, posB: Int = 0) {

        if (lastTopCounter != null) {
            val topDelta = calculateDelta(lastTopCounter!!, topCounter)
            if (topDelta > 0) {
                recordTopPosition(posA, posB)

                // SAFETY FIX: Stop at top position AFTER completing target reps
                // This ensures full tension through both concentric and eccentric of final rep
                // Changed from (workingTarget - 1) to (workingTarget) so we don't release early
                if (stopAtTop && !isJustLift && workingTarget > 0 && workingReps >= workingTarget) {
                    shouldStop = true
                    onRepEvent?.invoke(
                        RepEvent(
                            type = RepType.WORKOUT_COMPLETE,
                            warmupCount = warmupReps,
                            workingCount = workingReps
                        )
                    )
                }
            }
        }
        lastTopCounter = topCounter

        if (lastCompleteCounter == null) {
            lastCompleteCounter = completeCounter
            return  // Skip first signal to establish baseline
        }

        val delta = calculateDelta(lastCompleteCounter!!, completeCounter)
        if (delta <= 0) {
            return
        }

        lastCompleteCounter = completeCounter

        recordBottomPosition(posA, posB)

        val totalReps = warmupReps + workingReps + 1
        if (totalReps <= warmupTarget) {
            warmupReps++
            onRepEvent?.invoke(
                RepEvent(
                    type = RepType.WARMUP_COMPLETED,
                    warmupCount = warmupReps,
                    workingCount = workingReps
                )
            )
            if (warmupReps == warmupTarget) {
                onRepEvent?.invoke(
                    RepEvent(
                        type = RepType.WARMUP_COMPLETE,
                        warmupCount = warmupReps,
                        workingCount = workingReps
                    )
                )
            }
        } else {
            workingReps++
            onRepEvent?.invoke(
                RepEvent(
                    type = RepType.WORKING_COMPLETED,
                    warmupCount = warmupReps,
                    workingCount = workingReps
                )
            )

            // Only stop at bottom if stopAtTop is disabled
            // Most users should use stopAtTop (safer) but this preserves old behavior for those who want it
            if (!stopAtTop && !isJustLift && workingTarget > 0 && workingReps >= workingTarget) {
                shouldStop = true
                onRepEvent?.invoke(
                    RepEvent(
                        type = RepType.WORKOUT_COMPLETE,
                        warmupCount = warmupReps,
                        workingCount = workingReps
                    )
                )
            }
        }
    }

    private fun calculateDelta(last: Int, current: Int): Int {
        return if (current >= last) {
            current - last
        } else {
            0xFFFF - last + current + 1
        }
    }

    private fun recordTopPosition(posA: Int, posB: Int) {
        if (posA <= 0 && posB <= 0) return

        val window = getWindowSize()
        if (posA > 0) {
            topPositionsA.add(posA)
            if (topPositionsA.size > window) topPositionsA.removeAt(0)
        }
        if (posB > 0) {
            topPositionsB.add(posB)
            if (topPositionsB.size > window) topPositionsB.removeAt(0)
        }

        updateRepRanges()
    }

    private fun recordBottomPosition(posA: Int, posB: Int) {
        if (posA <= 0 && posB <= 0) return

        val window = getWindowSize()
        if (posA > 0) {
            bottomPositionsA.add(posA)
            if (bottomPositionsA.size > window) bottomPositionsA.removeAt(0)
        }
        if (posB > 0) {
            bottomPositionsB.add(posB)
            if (bottomPositionsB.size > window) bottomPositionsB.removeAt(0)
        }

        updateRepRanges()
    }

    private fun updateRepRanges() {
        if (topPositionsA.isNotEmpty()) {
            maxRepPosA = topPositionsA.average().toInt()
            maxRepPosARange = Pair(topPositionsA.minOrNull() ?: 0, topPositionsA.maxOrNull() ?: 0)
        }
        if (bottomPositionsA.isNotEmpty()) {
            minRepPosA = bottomPositionsA.average().toInt()
            minRepPosARange = Pair(bottomPositionsA.minOrNull() ?: 0, bottomPositionsA.maxOrNull() ?: 0)
        }
        if (topPositionsB.isNotEmpty()) {
            maxRepPosB = topPositionsB.average().toInt()
            maxRepPosBRange = Pair(topPositionsB.minOrNull() ?: 0, topPositionsB.maxOrNull() ?: 0)
        }
        if (bottomPositionsB.isNotEmpty()) {
            minRepPosB = bottomPositionsB.average().toInt()
            minRepPosBRange = Pair(bottomPositionsB.minOrNull() ?: 0, bottomPositionsB.maxOrNull() ?: 0)
        }
    }

    private fun getWindowSize(): Int {
        val total = warmupReps + workingReps
        return if (total < warmupTarget) 2 else 3
    }

    fun getRepCount(): RepCount {
        val total = workingReps  // Exclude warm-up reps from total count
        return RepCount(
            warmupReps = warmupReps,
            workingReps = workingReps,
            totalReps = total,
            isWarmupComplete = warmupReps >= warmupTarget
        )
    }

    fun shouldStopWorkout(): Boolean = shouldStop

    fun getCalibratedTopPosition(): Int? = maxRepPosA

    fun getRepRanges(): RepRanges = RepRanges(
        minPosA = minRepPosA,
        maxPosA = maxRepPosA,
        minPosB = minRepPosB,
        maxPosB = maxRepPosB,
        minRangeA = minRepPosARange,
        maxRangeA = maxRepPosARange,
        minRangeB = minRepPosBRange,
        maxRangeB = maxRepPosBRange
    )

    fun hasMeaningfulRange(minRangeThreshold: Int = 50): Boolean {
        val rangeA = if (minRepPosA != null && maxRepPosA != null) maxRepPosA!! - minRepPosA!! else 0
        val rangeB = if (minRepPosB != null && maxRepPosB != null) maxRepPosB!! - minRepPosB!! else 0
        return rangeA > minRangeThreshold || rangeB > minRangeThreshold
    }

    fun isInDangerZone(posA: Int, posB: Int, minRangeThreshold: Int = 50): Boolean {
        val checkA = minRepPosA != null && maxRepPosA != null
        val checkB = minRepPosB != null && maxRepPosB != null
        if (!checkA && !checkB) return false

        var danger = false
        if (checkA) {
            val rangeA = maxRepPosA!! - minRepPosA!!
            if (rangeA > minRangeThreshold) {
                val thresholdA = minRepPosA!! + (rangeA * 0.05f).toInt()
                danger = danger || posA <= thresholdA
            }
        }
        if (checkB) {
            val rangeB = maxRepPosB!! - minRepPosB!!
            if (rangeB > minRangeThreshold) {
                val thresholdB = minRepPosB!! + (rangeB * 0.05f).toInt()
                danger = danger || posB <= thresholdB
            }
        }
        return danger
    }
}

/**
 * Snapshot of the discovered rep ranges for UI/diagnostics.
 */
data class RepRanges(
    val minPosA: Int?,
    val maxPosA: Int?,
    val minPosB: Int?,
    val maxPosB: Int?,
    val minRangeA: Pair<Int, Int>?,
    val maxRangeA: Pair<Int, Int>?,
    val minRangeB: Pair<Int, Int>?,
    val maxRangeB: Pair<Int, Int>?
) {
    val rangeA: Int?
        get() = if (minPosA != null && maxPosA != null) max(maxPosA!! - minPosA!!, 0) else null
    val rangeB: Int?
        get() = if (minPosB != null && maxPosB != null) max(maxPosB!! - minPosB!!, 0) else null
}
